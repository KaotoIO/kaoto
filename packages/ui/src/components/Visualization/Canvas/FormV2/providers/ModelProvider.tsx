import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
