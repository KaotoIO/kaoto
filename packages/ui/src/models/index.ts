export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
