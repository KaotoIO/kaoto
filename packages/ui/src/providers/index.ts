export * from './catalog-schema-loader.provider';
export * from './entities.provider';
