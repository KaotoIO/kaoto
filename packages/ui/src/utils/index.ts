export * from './camel-case-to-space';
export * from './camel-uri-helper';
export * from './catalog-schema-loader';
export * from './create-camel-properties-sorter';
export * from './event-notifier';
export * from './get-array-property';
export * from './get-custom-schema-from-kamelet';
export * from './get-field-groups';
export * from './get-required-properties-schema';
export * from './get-serialized-model';
export * from './get-user-updated-properties-schema';
export * from './get-value';
export * from './get-viznodes-from-graph';
export * from './init-visible-flows';
export * from './is-datamapper';
export * from './is-defined';
export * from './is-enum-type';
export * from './is-to-processor';
export * from './is-xslt-component';
export * from './join-path';
export * from './node-icon-resolver';
export * from './pipe-custom-schema';
export * from './promise-timeout';
export * from './resolve-ref-if-needed';
export * from './set-value';
export * from './update-kamelet-from-custom-schema';
export * from './weight-schema-properties';
