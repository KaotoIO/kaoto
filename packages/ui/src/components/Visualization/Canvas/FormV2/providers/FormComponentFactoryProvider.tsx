import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
