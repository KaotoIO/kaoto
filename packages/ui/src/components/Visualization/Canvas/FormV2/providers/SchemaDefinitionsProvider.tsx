import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
