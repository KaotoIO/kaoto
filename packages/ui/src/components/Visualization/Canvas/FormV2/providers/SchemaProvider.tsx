import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
