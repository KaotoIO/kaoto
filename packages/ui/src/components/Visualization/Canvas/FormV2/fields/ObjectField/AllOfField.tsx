import { FunctionComponent, useContext } from 'react';
import { SchemaContext, SchemaProvider } from '../../providers/SchemaProvider';
import { FieldProps } from '../../typings';
import { AutoField } from '../AutoField';

export const AllOfField: FunctionComponent<FieldProps> = ({ propName, required, onRemove }) => {
  const { schema } = useContext(SchemaContext);
  if (!Array.isArray(schema.allOf)) {
    throw new Error('AllOfField: allOf must be an array');
  }

  return (
    <>
      {schema.allOf.map((schema, index) => {
        return (
          <SchemaProvider key={index} schema={schema}>
            <AutoField propName={propName} required={required} onRemove={onRemove} />
          </SchemaProvider>
        );
      })}
    </>
  );
};
