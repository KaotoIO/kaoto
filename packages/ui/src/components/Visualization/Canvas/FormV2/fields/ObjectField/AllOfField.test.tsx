import { render } from '@testing-library/react';
import { inspect } from 'node:util';
import { useContext } from 'react';
import { KaotoSchemaDefinition } from '../../../../../../models';
import { ROOT_PATH } from '../../../../../../utils';
import { FormComponentFactoryContext } from '../../providers/FormComponentFactoryProvider';
import { SchemaContext, SchemaProvider } from '../../providers/SchemaProvider';
import { FormWrapper } from '../../testing/FormWrapper';
import { AllOfField } from './AllOfField';

describe('AllOfField', () => {
  const allOf: KaotoSchemaDefinition['schema']['anyOf'] = [
    { type: 'object', properties: { name: { type: 'string', title: 'Name' } } },
    { type: 'object', properties: { valid: { type: 'boolean', title: 'Valid' } } },
  ];

  it('should render the entire anyOf schemas', () => {
    const wrapper = render(
      <SchemaProvider schema={{ allOf }}>
        <AllOfField propName={ROOT_PATH} />
      </SchemaProvider>,
      { wrapper: FormWrapper },
    );

    const inputFields = wrapper.queryAllByRole('textbox');
    expect(inputFields).toHaveLength(1);
    expect(inputFields[0]).toHaveAttribute('name', '#.name');

    const checkboxFields = wrapper.queryAllByRole('checkbox');
    expect(checkboxFields).toHaveLength(1);
    expect(checkboxFields[0]).toHaveAttribute('name', '#.valid');
  });

  it('should set the property name and propagate the required status', () => {
    const factorySpy = jest.fn().mockReturnValue((props: Record<string, string>) => {
      const { schema } = useContext(SchemaContext);

      return (
        <>
          <article>
            <h1>Schema</h1>
            <section>{inspect(schema)}</section>
          </article>

          <article>
            <h1>Props</h1>
            <section>{inspect(props)}</section>
          </article>
        </>
      );
    });

    const wrapper = render(
      <FormComponentFactoryContext.Provider value={factorySpy}>
        <SchemaProvider schema={{ allOf }}>
          <AllOfField propName={ROOT_PATH} />
        </SchemaProvider>
      </FormComponentFactoryContext.Provider>,
      { wrapper: FormWrapper },
    );

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
