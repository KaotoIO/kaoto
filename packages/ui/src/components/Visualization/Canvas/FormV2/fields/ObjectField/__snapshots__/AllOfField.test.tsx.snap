// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllOfField should set the property name and propagate the required status 1`] = `
<DocumentFragment>
  <article>
    <h1>
      Schema
    </h1>
    <section>
      {
  type: 'object',
  properties: { name: { type: 'string', title: 'Name' } }
}
    </section>
  </article>
  <article>
    <h1>
      Props
    </h1>
    <section>
      { propName: '#', required: undefined, onRemove: undefined }
    </section>
  </article>
  <article>
    <h1>
      Schema
    </h1>
    <section>
      {
  type: 'object',
  properties: { valid: { type: 'boolean', title: 'Valid' } }
}
    </section>
  </article>
  <article>
    <h1>
      Props
    </h1>
    <section>
      { propName: '#', required: undefined, onRemove: undefined }
    </section>
  </article>
</DocumentFragment>
`;
